### ISSUE TITLE
PropertiesExtension._get_property return types do not always correspond to `typ`

### ISSUE THREAD

**Issue by l0b0** [stac-utils/pystac#393]
While looking into how to make the mypy checks stricter I tried to add type assertions to this method:

```
def _get_property(self, prop_name: str, typ: Type[P]) -> Optional[P]:
    result = self.properties.get(prop_name)
    if result is not None:
        assert isinstance(result, typ), f"Type mismatch: {result} is not a {typ}"
        return result
    if self.additional_read_properties is not None:
        for props in self.additional_read_properties:
            result = props.get(prop_name)
            if result is not None:
                assert isinstance(
                    result, typ
                ), f"Type mismatch: {result} is not a {typ}"
                return result
    return None
```

This resulted in some test failures:

```
FAIL: test_cloud_cover (extensions.test_eo.EOTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/victor/dev/pystac/tests/extensions/test_eo.py", line 146, in test_cloud_cover
    cloud_cover = EOExtension.ext(item).cloud_cover
  File "/home/victor/dev/pystac/pystac/extensions/eo.py", line 332, in cloud_cover
    return self._get_property(CLOUD_COVER_PROP, float)
  File "/home/victor/dev/pystac/pystac/extensions/base.py", line 63, in _get_property
    assert isinstance(result, typ), f"Type mismatch: {result} is not a {typ}"
AssertionError: Type mismatch: 78 is not a <class 'float'>

======================================================================
FAIL: test_item_apply (extensions.test_eo.EOTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/victor/dev/pystac/tests/extensions/test_eo.py", line 228, in test_item_apply
    self.assertEqual(eo_ext.cloud_cover, 78)
  File "/home/victor/dev/pystac/pystac/extensions/eo.py", line 332, in cloud_cover
    return self._get_property(CLOUD_COVER_PROP, float)
  File "/home/victor/dev/pystac/pystac/extensions/base.py", line 63, in _get_property
    assert isinstance(result, typ), f"Type mismatch: {result} is not a {typ}"
AssertionError: Type mismatch: 78 is not a <class 'float'>

======================================================================
FAIL: test_density (extensions.test_pointcloud.PointcloudTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/victor/dev/pystac/tests/extensions/test_pointcloud.py", line 157, in test_density
    pc_density = PointcloudExtension.ext(pc_item).density
  File "/home/victor/dev/pystac/pystac/extensions/pointcloud.py", line 496, in density
    return self._get_property(DENSITY_PROP, float)
  File "/home/victor/dev/pystac/pystac/extensions/base.py", line 63, in _get_property
    assert isinstance(result, typ), f"Type mismatch: {result} is not a {typ}"
AssertionError: Type mismatch: 0 is not a <class 'float'>

======================================================================
FAIL: test_incidence_angle (extensions.test_view.ViewTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/victor/dev/pystac/tests/extensions/test_view.py", line 82, in test_incidence_angle
    view_incidence_angle = ViewExtension.ext(view_item).incidence_angle
  File "/home/victor/dev/pystac/pystac/extensions/view.py", line 102, in incidence_angle
    return self._get_property(INCIDENCE_ANGLE_PROP, float)
  File "/home/victor/dev/pystac/pystac/extensions/base.py", line 63, in _get_property
    assert isinstance(result, typ), f"Type mismatch: {result} is not a {typ}"
AssertionError: Type mismatch: 0 is not a <class 'float'>

======================================================================
FAIL: test_off_nadir (extensions.test_view.ViewTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/victor/dev/pystac/tests/extensions/test_view.py", line 48, in test_off_nadir
    view_off_nadir = ViewExtension.ext(view_item).off_nadir
  File "/home/victor/dev/pystac/pystac/extensions/view.py", line 87, in off_nadir
    return self._get_property(OFF_NADIR_PROP, float)
  File "/home/victor/dev/pystac/pystac/extensions/base.py", line 63, in _get_property
    assert isinstance(result, typ), f"Type mismatch: {result} is not a {typ}"
AssertionError: Type mismatch: 0 is not a <class 'float'>

======================================================================
FAIL: test_migrates_added_extension (serialization.test_migrate.MigrateTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/victor/dev/pystac/tests/serialization/test_migrate.py", line 80, in test_migrates_added_extension
    self.assertEqual(view_ext.off_nadir, 1)
  File "/home/victor/dev/pystac/pystac/extensions/view.py", line 87, in off_nadir
    return self._get_property(OFF_NADIR_PROP, float)
  File "/home/victor/dev/pystac/pystac/extensions/base.py", line 63, in _get_property
    assert isinstance(result, typ), f"Type mismatch: {result} is not a {typ}"
AssertionError: Type mismatch: 1 is not a <class 'float'>

----------------------------------------------------------------------
Ran 305 tests in 50.397s

FAILED (failures=6, errors=26)
```

These tests should probably use floats for the test data.

(The errors above are instances of "TypeError: Subscripted generics cannot be used with class and instance checks", which just means my `isinstance` hack isn't appropriate for generics, and [libraries might be necessary to do such checks](https://stackoverflow.com/q/51171908/96588).)

**Comment by duckontheweb** [comment]
@l0b0 It seems like maybe this was fixed already by 01c924a9de5dbef23ca82e6c9516b01351a6d752?

**Comment by l0b0** [comment]
It doesn't look like it. If you apply the following diff and run the tests, a bunch of them start failing:

```patch
diff --git a/pystac/extensions/base.py b/pystac/extensions/base.py
index d33707f..5a9cad8 100644
--- a/pystac/extensions/base.py
+++ b/pystac/extensions/base.py
@@ -60,11 +60,15 @@ class PropertiesExtension(ABC):
     def _get_property(self, prop_name: str, typ: Type[P]) -> Optional[P]:
         maybe_property: Optional[P] = self.properties.get(prop_name)
         if maybe_property is not None:
+            assert isinstance(maybe_property, typ), f"Type mismatch: {maybe_property} is not a {typ}"
             return maybe_property
         if self.additional_read_properties is not None:
             for props in self.additional_read_properties:
                 maybe_additional_property: Optional[P] = props.get(prop_name)
                 if maybe_additional_property is not None:
+                    assert isinstance(
+                        maybe_additional_property, typ
+                    ), f"Type mismatch: {maybe_additional_property} is not a {typ}"
                     return maybe_additional_property
         return None
 
```
